src/Elastic.Apm.Azure.Storage/AzureFileShareStorageDiagnosticListener.cs

```diff
@@ -0,0 +1,186 @@
+﻿// Licensed to Elasticsearch B.V under
+// one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Elastic.Apm.Api;
+using Elastic.Apm.DiagnosticListeners;
+using Elastic.Apm.Logging;
+
+namespace Elastic.Apm.Azure.Storage
+{
+	internal static class AzureFileStorage
+	{
+		internal const string SpanName = "AzureFile";
+		internal const string SubType = "azurefile";
+	}
+
+	/// <summary>
+	/// Creates transactions and spans for Azure File Share Storage diagnostic events from Azure.Storage.Files.Shares
+	/// </summary>
+	internal class AzureFileShareStorageDiagnosticListener : DiagnosticListenerBase
+	{
+		private readonly ConcurrentDictionary<string, IExecutionSegment> _processingSegments =
+			new ConcurrentDictionary<string, IExecutionSegment>();
+
+		public AzureFileShareStorageDiagnosticListener(IApmAgent agent) : base(agent) { }
+
+		public override string Name { get; } = "Azure.Storage.Files.Shares";
+
+		protected override void HandleOnNext(KeyValuePair<string, object> kv)
+		{
+			Logger.Trace()?.Log("Called with key: `{DiagnosticEventKey}'", kv.Key);

```

Comment:
A bit off-topic comment, but maybe we could move this (and the same logs from the other new listeners) into `DiagnosticListenerBase` - we do this logging in lots of types that derive from `DiagnosticListenerBase` - we could do this once in the base class.

Doesn't have to be in this PR - this was just a thought when I looked at this.

Reply:
agreed, be good to move to the base class. Will open a follow up PR to do so.